sourced/ml/transformers/basic.py

```diff
@@ -32,10 +32,10 @@ def __call__(self, head: RDD):
         # partitionBy the key extracted using self.keymap
         if self.keymap is not False:
             # user knows what they are doing
-            head = head.map(lambda x: (self.keymap(x), x))
-        return head \
-            .partitionBy(self.partitions) \
-            .map(lambda x: x[1])
+            return head \
+                .map(lambda x: (self.keymap(x), x)) \
+                .partitionBy(self.partitions) \
+                .map(lambda x: x[1])
```

Comment:
So if `self.keymap is False` we return nothing.
None is not the same as False. But damn the whole logic here is twisted. Can ask the code author what we should do here and also add the comment which explains that `keymap` can be False, None or real callable.
Personally, I think that having a difference between False and None is very confusing.

Reply:
Yeah, I agree it is very confusing. But there is a comment in class __docstring__:
```python
"""
     Repartitioner uses one of the three ways to split an RDD into partitions:
     1. coalesce() if shuffle=False, keymap=None
     2. repartition() if shuffle=True, keymap=None
     3. partitionBy() if keymap is not None
"""
```

I think we can separate 3rd case into separate Transformer.

Reply:
No need to separate and the comment is good.

The confusion comes from different actions taken when keymap is None and keymap is False.

Reply:
ok, now I get it. @warenlg now I noticed that you changed the logic of this code. 
Now it is
```python
    def __call__(self, head: RDD):
        if self.keymap is None:
            return head.coalesce(self.partitions, self.shuffle)
        # partitionBy the key extracted using self.keymap
        if self.keymap is not False:
            # user knows what they are doing
            return head \
                .map(lambda x: (self.keymap(x), x)) \
                .partitionBy(self.partitions) \
                .map(lambda x: x[1])
```
and it was
```python
    def __call__(self, head: RDD):
        if self.keymap is None:
            return head.coalesce(self.partitions, self.shuffle)
        # partitionBy the key extracted using self.keymap
        if self.keymap is not False:
            # user knows what they are doing
            head = head.map(lambda x: (self.keymap(x), x))
        return head \
            .partitionBy(self.partitions) \
            .map(lambda x: x[1])
```

So in case `self.keymap is False` you return nothing and it is an error. if `self.keymap is False` transformer should just skip mapping step.
I think it is good idea to cover this option with additional test.